### PR TITLE
CLI: Skip vitest transform for CSF Factories in a11y-addon-test automigration

### LINKED ISSUE #31940: [Bug]: `addon-a11y-addon-test` migration fails if using csf factories
### Describe the bug

Running the above migration against a storybook using csf factories fails:

```
  1) We couldn't find or automatically update .storybook/vitest.setup.<ts|js> in your project to smoothly set up     │
│   project annotations from @storybook/addon-a11y.                                                                    │
│   Please manually update your vitest.setup.ts file to include the following:                                         │
│                                                                                                                      │
│   ...                                                                                                                │
│   + import * as a11yAddonAnnotations from "@storybook/addon-a11y/preview";                                           │
│                                                                                                                      │
│   setProjectAnnotations([                                                                                            │
│   ...                                                                                                                │
│   + a11yAddonAnnotations,                                                                                            │
│   ]);                      
```
unsurprisingly, as the above snippet is irrelevant to csf factories

### Reproduction link

-

### Reproduction steps

_No response_

### System

```bash
-
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mrginglymus**
Closes #31940

## What I did

Sniff for usage of csf factories in `vitest.setup` and skip transformation

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [x] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli-storybook/src/sandbox-templates.ts`
- [x] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

  - `bug`: Internal changes that fixes incorrect behavior.
  - `maintenance`: User-facing maintenance tasks.
  - `dependencies`: Upgrading (sometimes downgrading) dependencies.
  - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
  - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
  - `documentation`: Documentation **only** changes. Will not show up in release changelog.
  - `feature request`: Introducing a new feature.
  - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
  - `other`: Changes that don't fit in the above categories.

   </details>

### 🦋 Canary release

This pull request has been released as version `0.0.0-pr-31941-sha-498c84dc`. Try it out in a new sandbox by running `npx storybook@0.0.0-pr-31941-sha-498c84dc sandbox` or in an existing project with `npx storybook@0.0.0-pr-31941-sha-498c84dc upgrade`.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-31941-sha-498c84dc`](https://npmjs.com/package/storybook/v/0.0.0-pr-31941-sha-498c84dc) |
| **Triggered by** | @valentinpalkovic |
| **Repository** | [mrginglymus/storybook](https://github.com/mrginglymus/storybook) |
| **Branch** | [`addon-a11y-migrate-fix`](https://github.com/mrginglymus/storybook/tree/addon-a11y-migrate-fix) |
| **Commit** | [`498c84dc`](https://github.com/mrginglymus/storybook/commit/498c84dc383ca352e62a21de534751de96ba2340) |
| **Datetime** | Wed Dec  3 12:03:03 UTC 2025 (`1764763383`) |
| **Workflow run** | [19893145747](https://github.com/storybookjs/storybook/actions/runs/19893145747) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/publish.yml) or locally with `gh workflow run --repo storybookjs/storybook publish.yml --field pr=31941`_
</details>

## Greptile Summary

Modified the a11y addon migration logic to prevent incorrect transformations of `vitest.setup` files when CSF factories are detected. This fixes an issue where the standard transformation would fail for projects using Component Story Format factories.

- Added detection for `preview.composed.beforeAll` in setup files to identify CSF factory usage
- Updated `addonA11yAddonTest.check()` to set `skipVitestSetupTransformation=true` when factories are detected
- Added test coverage verifying the detection and handling of CSF factory scenarios
- Addresses incompatibility issues between the default migration path and CSF factory setups

## Summary by CodeRabbit

- New Features
  - Automigrations now detect CSF 4 usage and tailor fixes accordingly.
  - Improved compatibility with projects using CSF factory previews.

- Bug Fixes
  - More accurate handling of a11y addon test migration under CSF 4.
  - Avoids unnecessary Vitest setup transformation when CSF 4 is detected.

### CODE REVIEW

File: code/lib/cli-storybook/src/automigrate/fixes/addon-a11y-addon-test.ts [hunk 1]
```diff
@@ -78,7 +78,9 @@ export const addonA11yAddonTest: Fix<AddonA11yAddonTestOptions> = {
       const vitestSetupSource = readFileSync(vitestSetupFile, 'utf8');
       const previewSetupSource = readFileSync(previewFile, 'utf8');
 
-      skipVitestSetupTransformation = vitestSetupSource.includes('@storybook/addon-a11y');
+      skipVitestSetupTransformation =
+        vitestSetupSource.includes('@storybook/addon-a11y') ||
+        vitestSetupSource.includes('preview.composed.beforeAll');
```
Comment (valentinpalkovic) [inline code/lib/cli-storybook/src/automigrate/fixes/addon-a11y-addon-test.ts]: It makes more sense to pass a flag into each automigration instead and define the logic to evaluate whether csf4 is used or not at a higher level. Would you like to work on this?

Comment (mrginglymus) [reply code/lib/cli-storybook/src/automigrate/fixes/addon-a11y-addon-test.ts]: Presumably this would come from sniffing for `definePreview` in the detected preview file, around here: https://github.com/storybookjs/storybook/blob/3ac2fece4c41955e7349f6f825b7d21dd18ff9a9/code/lib/cli-storybook/src/automigrate/helpers/mainConfigFile.ts#L127 ?

I'd be happy to have a crack at this, but it would be ideal to unblock automigrations with this fix for now as it might be a few days before I can do it justice.

Comment (mrginglymus) [reply code/lib/cli-storybook/src/automigrate/fixes/addon-a11y-addon-test.ts]: Hey, sorry, I completelyy forgot about this. I'll see if I can get something whipped up today.

Comment (mrginglymus) [reply code/lib/cli-storybook/src/automigrate/fixes/addon-a11y-addon-test.ts]: Hmm, taking a look...it'll require a fair rewrite of addon testing to do this fairly. All the tests call `check` or `run` directly, so we won't be able to actually the logic in `getStorybookData`.

Comment (Sidnioulz) [reply code/lib/cli-storybook/src/automigrate/fixes/addon-a11y-addon-test.ts]: You could workaround the difficulty of defining your check by making this check (and possibly the code that identifies the vitest setup file? I'll let you be the judge of that) into `code/lib/cli-storybook/src/automigrate/helpers/testing-helpers.ts`.

This way, it'll be easier to reuse across migrations and you don't have to figure out how to bend more internals to your will.

[the remark this answers is not in the document]
Comment (Sidnioulz) [reply code/lib/cli-storybook/src/automigrate/fixes/addon-a11y-addon-test.ts]: ~Greptile is right, this would need to be documented. Is `preview.composed` a sign of CSF4? Is `beforeAll` bound to be defined, or is it only defined if addons that use it are installed?~

I get it now: you're looking for evidence that the vitest setup file is consuming `preview`'s `composed` property. I would prefer if this was done by AST: identify the preview import's identity, and check whether it's `composed.beforeAll` property is well mapped.

This would make for a more robust migration and would help us identify (and fix?) use cases where the vitest setup file is not well formed.

**Comment by Sidnioulz** [comment]
@mrginglymus could you please let me know how feasible it'd be to use an AST-based approach to detect preview? We have a few examples using Babel's AST tools in other migrations.

Also, in https://github.com/storybookjs/storybook/pull/32083, I set up a test bed with the few scenarios/edge cases to handle spread over branches. This immensely speeds up testing and review, as reviewers otherwise have to build a sandbox from scratch for every iteration of the PR. It would be lovely (but not required :) ) if you could set one up too.

**Comment by mrginglymus** [comment]
@Sidnioulz I've pushed the AST approach - there was already a utility for csf4 sniffing so I've used that in the project gathering step and propagated it to all migrations. 

That does mean we have no automated tests now because the migration tests only test code _after_ the gather phase; I'll need to sniff around for tests that capture an earlier phase - or, as you say, set up a repro.

**Comment by Sidnioulz** [comment]
> That does mean we have no automated tests now because the migration tests only test code _after_ the gather phase; I'll need to sniff around for tests that capture an earlier phase - or, as you say, set up a repro.

I feel that should be acceptable. You could use the SB logger and add debug messages in your migration code, or even assertions that would fail when your script finds an illogical/incoherent situation.

If your script fails when it can't produce a coherent output, then users know when to switch to manual migration. Part of the deal is we need to point to the right documentation resource for them.

Besides, if we have an E2E test somewhere, then everything works at the time the script is most needed. It might break later (as NPM deps evolve), and at the moment, we don't run any repro case in CI for any automigration. It might be something we'd want to automate, but the ROI isn't as clear or high as other things the team is working on, so it hasn't happened yet.

File: code/lib/cli-storybook/src/automigrate/helpers/mainConfigFile.ts [hunk 2]
```diff
@@ -184,6 +191,7 @@ export const getStorybookData = async ({
     previewConfigPath,
     packageManager,
     storiesPaths,
+    csf4,
```
Comment (Sidnioulz) [inline code/lib/cli-storybook/src/automigrate/helpers/mainConfigFile.ts]: nit: I think this could be named `hasCSF4Preview` or `hasCSFFactory` instead. @kasperpeulen might have an opinion as to what the best naming is.

It's possible to have mixed CSF3/CSF4 stories, and I wouldn't want us to start using this variable as evidence that a story file being automigrated is in CSF4 syntax, as the function specifically checks that the preview file is a CSF factory.

**Comment by Sidnioulz** [comment]
@mrginglymus I've updated your branch to trigger a fresh Windows test build as I think it was outdated, but FYI there are TS errors showing up in the CircleCI jobs, and you have failing tests for your automigration too.

See https://app.circleci.com/pipelines/github/storybookjs/storybook/104084/workflows/b6438da6-cd7b-4215-8684-e8bcefb1901f

**Review by Sidnioulz — APPROVED** [review]

**Comment by Sidnioulz** [comment]
You didn't provide manual testing instructions, but I trust you have tested against a real use case @mrginglymus :)

**Comment by mrginglymus** [comment]
Actually I could do with a canary for this one...

**Comment by Sidnioulz** [comment]
> Actually I could do with a canary for this one...

Valentin triggered the canary creation, it should appear in your PR description shortly!

**Comment by mrginglymus** [comment]
Sad trombone, it doesn't work.

**Comment by Sidnioulz** [comment]
> Sad trombone, it doesn't work.

Do you have leads to investigate, or would you need some time with the team to talk it out?

**Comment by mrginglymus** [comment]
> > Sad trombone, it doesn't work.
> 
> Do you have leads to investigate, or would you need some time with the team to talk it out?

Simple logic bug, but the test case wasn't realistic enough to capture it. Fixed now.

**Review by Sidnioulz — APPROVED** [review]
